src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java

```diff
@@ -0,0 +1,135 @@
+      "DejetifySrcs",
+      "Desugar",
+      "DexBuilder",
+      "Javac",
+      "JavacTurbine",
+      "Jetify",
+      "JetifySrcs",
+      "LinkAndroidResources",
+      "MergeAndroidAssets",
+      "MergeManifests",
+      "ParseAndroidResources",
+      "StarlarkAARGenerator",
+      "StarlarkMergeCompiledAndroidResources",
+      "StarlarkRClassGenerator",
+      "Turbine"
+  );
+
+  /**
+   * Actions that support path mapping should call this method from {@link
+   * Action#getKey(ActionKeyContext, ArtifactExpander)}.
+   *
+   * <p>Compared to {@link #create(Action, OutputPathsMode)}, this method does not flatten nested
+   * sets and thus can't result in memory regressions.
+   *
+   * @param mnemonic the mnemonic of the action
+   * @param outputPathsMode the value of {@link CoreOptions#outputPathsMode}
+   * @param fingerprint the fingerprint to add to
+   */
+  public static void addToFingerprint(String mnemonic, OutputPathsMode outputPathsMode,
+      Fingerprint fingerprint)
+      throws CommandLineExpansionException, InterruptedException {
+    // Creating a new PathMapper instance can be expensive, but isn't needed here: Whether and
+    // how path mapping applies to the action only depends on the output paths mode and the action
+    // inputs, which are already part of the action key.
+    OutputPathsMode effectiveOutputPathsMode = getEffectiveOutputPathsMode(outputPathsMode,
+        mnemonic);
+    if (effectiveOutputPathsMode == OutputPathsMode.STRIP) {
+      fingerprint.addString(StrippingPathMapper.GUID);
+    }
+  }
+
+  /**
+   * Actions that support path mapping should call this method when creating their {@link Spawn}.
+   *
+   * <p>The returned {@link PathMapper} has to be passed to {@link
+   * com.google.devtools.build.lib.actions.CommandLine#arguments(ArtifactExpander, PathMapper)},
+   * {@link com.google.devtools.build.lib.actions.CommandLines#expand(ArtifactExpander, PathFragment, PathMapper, CommandLineLimits)}
+   * or any other variants of these functions. The same instance should also be passed to the
+   * {@link Spawn} constructor so that the executor can obtain it via {@link Spawn#getPathMapper()}.
+   *
+   * <p>Note: This method flattens nested sets and should thus not be called from methods that are
+   * executed in the analysis phase.
+   *
+   * <p>Actions calling this method should also call {@link
+   * #addToFingerprint(String, OutputPathsMode, Fingerprint)} from {@link
+   * Action#getKey(ActionKeyContext, ArtifactExpander)} to ensure correct incremental builds.
+   *
+   * @param action the {@link Action} for which a {@link Spawn} is to be created
+   * @param outputPathsMode the value of {@link CoreOptions#outputPathsMode}
+   * @return a {@link PathMapper} that maps paths of the action's inputs and outputs. May be
+   *    {@link PathMapper#NOOP} if path mapping is not applicable to the action.
+   */
+  public static PathMapper create(Action action, OutputPathsMode outputPathsMode) {
```

Comment:
sanity check, since PathMapper is created during the Spawn creation, is `spawn.getPathMapper` the proper replacement for the previous `PathMappers.createForExecutor`

Reply:
Yes, exactly!